cryptoki/src/types/object.rs

```diff
@@ -430,11 +858,49 @@ impl TryFrom<CK_ATTRIBUTE> for Attribute {
                     .collect::<Result<Vec<MechanismType>>>()?;
                 Ok(Attribute::AllowedMechanisms(types))
             }
+            AttributeType::EndDate => {
+                let date = val.as_ptr() as *const CK_DATE;
+                unsafe {
+                    let year = String::from_utf8_lossy(Vec::from((*date).year).as_slice())
+                        .trim_end()
+                        .to_string();
```

Comment:
Since this is in a `TryFrom`, we should probably return an error if this conversion fails instead of just returning a "buggy" date. I'm also wondering if in the `new_from_str_slice` we should do something similar where we at least check that every character is a digit.

Reply:
I've been going back and forth on how to answer this.  I think it comes down to how much 'checking' a PKCS11 library does vs the token and what the user _intends_ the date to mean.  Frankly, the `CK_DATE` type is more annoying than it's worth.  Its only use is for "reference only" and the token can safely ignore the values within it.

I think from a "helping the user" perspective, it would be helpful to bounds check the date _when the user provides it_.  The issue I think comes down to the fact that the library does not have any control over the module not providing back a valid date for an object (e.g. someone creates a key with a `CKA_END_DATE` of 99/99/9999 or maybe NO/EX/PIRE to indicate that the key never expires.  Both values are technically valid).  That may not be a _valid_ date, but it is a valid 'value' that can be held by `CK_DATE`.  The value of a `CK_DATE` is up to the application to set and interpret.

The spec states that you should only use the characters in "Table 3" !["Table 3"](https://user-images.githubusercontent.com/3419731/133109921-914ef88f-d1e7-4374-9a82-e396de2b5192.png) for any `CK_CHAR` type.  `CK_UTF8CHAR` is governed by ![RFC2279](https://datatracker.ietf.org/doc/html/rfc2279).  Meaning as long as the values of a `CK_CHAR` field are within the bounds stated by Table 3, it's valid.

Thoughts?

Reply:
In that case let's just leave it as is, it doesn't seem like something critical anyway. We can implement a method on `Date` to check whether it's valid or not, if someone requests it.